core/src/modules/ops/source.ts

```diff
@@ -253,6 +258,12 @@ export namespace SourceOps {
         profileIds: profiles.map((p) => p.id),
       });
 
+      await applyNonUniqueMappedResultsToAllProfiles(response, {
+        profiles,
+        properties,
+        sourceMapping,
+      });
+
```

Comment:
I understand that this new method sets the non-uniquely mapped properties values for all other pending profiles, but it's not totally clear to me why this is necessary. Does the `profileProperties` method not already return a response with the values set for each of the profiles in question?

Looking a bit into the `app-template`'s `profileProperties` method implementation, it seems like if the "primaryKey" (the mapping) matches a profile, it will be pushed to an array:
https://github.com/grouparoo/grouparoo/blob/b31360b3418a0a7b9f04b454d9a336d9ca1b6e55/plugins/%40grouparoo/app-templates/src/source/table/profileProperties.ts#L64-L66

...and then the resulting value will be set on each of those profiles:
https://github.com/grouparoo/grouparoo/blob/b31360b3418a0a7b9f04b454d9a336d9ca1b6e55/plugins/%40grouparoo/app-templates/src/source/table/profileProperties.ts#L110-L120

From what I'm reading, it suggests that multiple profiles can have this same mapping value and it should set it on all, no? What am I missing?

Reply:
> I understand that this new method sets the non-uniquely mapped properties values for all other pending profiles, but it's not totally clear to me why this is necessary. Does the `profileProperties` method not already return a response with the values set for each of the profiles in question?
> ...
> From what I'm reading, it suggests that multiple profiles can have this same mapping value and it should set it on all, no? What am I missing?


Yeah, this was the confusing part!

https://github.com/grouparoo/grouparoo/blob/b31360b3418a0a7b9f04b454d9a336d9ca1b6e55/plugins/%40grouparoo/app-templates/src/source/table/profileProperties.ts#L64-L66 sets up the response object to have an entry for each profile.  

https://github.com/grouparoo/grouparoo/blob/b31360b3418a0a7b9f04b454d9a336d9ca1b6e55/plugins/%40grouparoo/app-templates/src/source/table/profileProperties.ts#L110-L120 takes what the plugin returns and applies to the response hash per profile, per property... BUT it doesn't guarantee that every profile sent to the plugin will get a result back in this case.

I've been using a translation [dataset in my tests](https://github.com/grouparoo/grouparoo/blob/f06019c64299b7e900fa3ae4e7ddfbe49b875798/core/__tests__/data/translations.csv).  Imagine that was SQL, and we had 100 Profiles with the property `english_word="Spanish"` and `english_word="English"` and we were trying to load up the property `spanish_word`.  The generated query would be something like:
  
```sql
SELECT "spanish_word", "english_word" as __pk FROM translations WHERE  "english_word" IN ('Spanish','English') GROUP BY english_word,spanish_word
```
... which means we'll only get one row back from the database for each language here (2 rows total), not the 100 we need.  So, the code then "fans out" (eg: copies) the response over and over to all the Profiles that has the same `english_word`.

The alternative would have been to come up with a query like @tealjulia talks about in her [0's blog post](https://www.grouparoo.com/blog/data-cleaning-null-aggregate-values) to union together a result set that duplicates the data in SQL rather than JS.  I chose the in-code way to make this more generic for all sources.